File: swift/DesignPatterns/Creational/Builder.swift
//
//  Builder.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 06/06/2026.
//

import Foundation

// Without Builder — plain constructor (initializer) approach.
// Every field must be passed in one long call, and at the call site
// it is hard to tell which argument is which in languages without
// argument labels (like Java):
//
// struct Person {
//     let firstName: String
//     let lastName: String
//     let age: Int
//     let email: String
//     let phone: String
//     let address: String
//
//     init(firstName: String, lastName: String, age: Int,
//          email: String, phone: String, address: String) {
//         self.firstName = firstName
//         self.lastName = lastName
//         self.age = age
//         self.email = email
//         self.phone = phone
//         self.address = address
//     }
// }
//
// let person = Person(firstName: "John", lastName: "Doe", age: 30,
//                     email: "john.doe@example.com",
//                     phone: "+1234567890",
//                     address: "123 Main Street")

struct Person: CustomStringConvertible {
    let firstName: String
    let lastName: String
    let age: Int?
    let email: String?
    let phone: String?
    let address: String?
    
    private init(builder: Builder) {
        self.firstName = builder.firstName
        self.lastName = builder.lastName
        self.age = builder.age
        self.email = builder.email
        self.phone = builder.phone
        self.address = builder.address
    }
    
    var description: String {
        "Person{firstName='\(firstName)', lastName='\(lastName)', age=\(age.map(String.init) ?? "nil"), email='\(email ?? "nil")', phone='\(phone ?? "nil")', address='\(address ?? "nil")'}"
    }
    
    class Builder {
        fileprivate let firstName: String
        fileprivate let lastName: String
        fileprivate var age: Int?
        fileprivate var email: String?
        fileprivate var phone: String?
        fileprivate var address: String?
        
        init(firstName: String, lastName: String) {
            self.firstName = firstName
            self.lastName = lastName
        }
        
        @discardableResult
        func age(_ age: Int) -> Builder {
            self.age = age
            return self
        }
        
        @discardableResult
        func email(_ email: String) -> Builder {
            self.email = email
            return self
        }
        
        @discardableResult
        func phone(_ phone: String) -> Builder {
            self.phone = phone
            return self
        }
        
        @discardableResult
        func address(_ address: String) -> Builder {
            self.address = address
            return self
        }
        
        func build() -> Person {
            Person(builder: self)
        }
    }
}

func builderExample() {
    let person = Person.Builder(firstName: "John", lastName: "Doe")
        .age(30)
        .email("john.doe@example.com")
        .phone("+1234567890")
        .address("123 Main Street")
        .build()
    print(person)
}

File: dart/design_patterns/lib/creational/builder.dart
// Without Builder — plain constructor approach.
// Every field must be passed in one long call:
//
// class Person {
//   final String firstName;
//   final String lastName;
//   final int age;
//   final String email;
//   final String phone;
//   final String address;
//
//   Person(this.firstName, this.lastName, this.age,
//       this.email, this.phone, this.address);
// }
//
// final person = Person('John', 'Doe', 30, 'john.doe@example.com',
//     '+1234567890', '123 Main Street');

// Product

class Person {
  final String firstName;
  final String lastName;
  final int? age;
  final String? email;
  final String? phone;
  final String? address;

  Person._(PersonBuilder builder)
      : firstName = builder._firstName,
        lastName = builder._lastName,
        age = builder._age,
        email = builder._email,
        phone = builder._phone,
        address = builder._address;

  @override
  String toString() {
    return "Person{firstName='$firstName', lastName='$lastName', "
        "age=$age, email='$email', phone='$phone', address='$address'}";
  }
}

// Builder

class PersonBuilder {
  final String _firstName;
  final String _lastName;
  int? _age;
  String? _email;
  String? _phone;
  String? _address;

  PersonBuilder(this._firstName, this._lastName);

  PersonBuilder age(int age) {
    _age = age;
    return this;
  }

  PersonBuilder email(String email) {
    _email = email;
    return this;
  }

  PersonBuilder phone(String phone) {
    _phone = phone;
    return this;
  }

  PersonBuilder address(String address) {
    _address = address;
    return this;
  }

  Person build() => Person._(this);
}

// Usage

void main() {
  final person = PersonBuilder('John', 'Doe')
      .age(30)
      .email('john.doe@example.com')
      .phone('+1234567890')
      .address('123 Main Street')
      .build();
  print(person);
}
